processor/spanprocessor/span.go

```diff
@@ -121,3 +161,76 @@ func (sp *spanProcessor) nameSpan(span *tracepb.Span) {
 	}
 	span.Name = &tracepb.TruncatableString{Value: sb.String()}
 }
+
+func (sp *spanProcessor) processToAttributes(span *tracepb.Span) {
+	if span.Name == nil || span.Name.Value == "" {
+		// There is no span name to work on.
+		return
+	}
+
+	if sp.config.Rename.ToAttributes == nil {
+		// No rules to apply.
+		return
+	}
+
+	// Process rules one by one. Store results of processing in the span
+	// so that each subsequent rule works on the span name that is the output
+	// after processing the previous rule.
+	for _, rule := range sp.toAttributeRules {
+		re := rule.re
+		oldName := span.Name.Value
+
+		// Match the regular expression and extract matched subexpressions.
+		if submatches := re.FindStringSubmatch(oldName); submatches != nil {
+			// There is a match. We will also need positions of subexpression matches.
+			submatchIdxs := re.FindStringSubmatchIndex(oldName)
+
+			// A place to accumulate new span name.
+			var sb strings.Builder
+
+			// Index in the oldName until which we traversed.
+			var oldNameIndex = 0
+
+			// Ensure we have a place to create attributes.
+			if span.Attributes == nil {
+				span.Attributes = &tracepb.Span_Attributes{}
+			}
+			attrs := span.Attributes
+
+			// Create a new map if one does not exist and size it to the number of
+			// attributes that we plan to extract.
+			if attrs.AttributeMap == nil {
+				attrs.AttributeMap = make(map[string]*tracepb.AttributeValue, len(rule.attrNames))
+			}
+
+			// Start from index 1, which is the first submatch (index 0 is the entire match).
+			// We will go over submatches and will simultaneously build a new span name,
+			// replacing matched subexpressions by attribute names.
+			for i := 1; i < len(submatches); i++ {
+				// Create a string attribute from extracted submatch.
+				attrValue := &tracepb.AttributeValue{Value: &tracepb.AttributeValue_StringValue{
+					StringValue: &tracepb.TruncatableString{Value: submatches[i]},
+				}}
+				attrs.AttributeMap[rule.attrNames[i]] = attrValue
+
+				// Add part of span name from end of previous match to start of this match
+				// and then add attribute name wrapped in curly brackets.
+				sb.WriteString(oldName[oldNameIndex:submatchIdxs[i*2]] + "{" + rule.attrNames[i] + "}")
```

Comment:
The use of `*2` is not clear to me, could you please add a comment about it? Likely describing the content of `submatchIdxs`.

Reply:
FindStringSubmatchIndex returns pairs of indexes for each match.

Reply:
Added temp var to make clearer and renamed to submatchIdxs to submatchIdxPairs.